irods/query2.py

```diff
@@ -0,0 +1,47 @@
+import json
+
+from irods.api_number import api_number
+from irods.exception import OperationNotSupported
+from irods.message import GenQuery2Request, STR_PI, iRODSMessage
+
+
+class Query2(object):
+    """This class provides an interface to GenQuery2 API, an experimental
+       iRODS API for querying iRODS. GenQuery2 is an improved version of the
+       traditional GenQuery interface.
+    """
+
+    def __init__(self, session, zone, query, override_not_supported=False):
+        self.session = session
+        self.zone = zone
+        self.query = query
+        if not (self._is_supported() or override_not_supported):
+            raise OperationNotSupported(
+                "GenQuery2 is not supported by default on this iRODS version.")
+
+    def execute(self):
+        """Execute this GenQuery2 query, and return the results."""
+        return self._exec_genquery2()
+
+    def get_sql(self):
+        """Return the SQL query that this GenQuery2 query will be translated to."""
+        return self._exec_genquery2(sql_flag=True)
+
+    def _exec_genquery2(self, sql_flag=False):
+        msg = GenQuery2Request()
+        msg.query_string = self.query
+        msg.zone = self.zone
+        msg.sql_only = 1 if sql_flag else 0
+        msg.column_mappings = 0
+        message = iRODSMessage('RODS_API_REQ',
+                               msg=msg,
+                               int_info=api_number['GENQUERY2_AN'])
+        with self.session.pool.get_connection() as conn:
+            conn.send(message)
+            response = conn.recv()
+        result = response.get_main_message(STR_PI).myStr
+        return result if sql_flag else json.loads(result)
+
+    def _is_supported(self):
+        """Checks whether this iRODS server supports GenQuery2."""
+        return self.session.server_version >= (4, 3, 2)
```

Comment:
For the future, we should consider adding a way to invoke the `get_library_features` endpoint in PRC and then use that to check for the GenQuery2 feature. I've created an issue for this here: https://github.com/irods/python-irodsclient/issues/556

Reply:
Just to be clear, I'm not asking @stsnel to implement this feature as part of this PR and in fact would prefer that it not be part of this PR. :) Just opening the discussion in case there is disagreement on the approach.

Reply:
That's exactly what the library features endpoint was designed for.

And I agree it isn't required of this PR.

Reply:
Cool, resolving.